homeassistant/components/eufylife_ble/sensor.py

```diff
@@ -0,0 +1,211 @@
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.restore_state import RestoreEntity
+from homeassistant.util.unit_conversion import MassConverter
+from homeassistant.util.unit_system import US_CUSTOMARY_SYSTEM
+
+from .const import DOMAIN
+from .models import EufyLifeData
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: config_entries.ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the EufyLife sensors."""
+    data: EufyLifeData = hass.data[DOMAIN][entry.entry_id]
+
+    entities = [
+        EufyLifeWeightSensorEntity(data),
+        EufyLifeRealTimeWeightSensorEntity(data),
+    ]
+
+    if data.client.supports_heart_rate:
+        entities.append(EufyLifeHeartRateSensorEntity(data))
+
+    async_add_entities(entities)
+
+
+class EufyLifeSensorEntity(SensorEntity):
+    """Representation of an EufyLife sensor."""
+
+    _attr_has_entity_name = True
+
+    def __init__(self, data: EufyLifeData) -> None:
+        """Initialize the weight sensor entity."""
+        self._data = data
+
+        self._attr_device_info = DeviceInfo(
+            name=MODEL_TO_NAME[data.model],
+            connections={(dr.CONNECTION_BLUETOOTH, data.address)},
+        )
+
+    @property
+    def available(self) -> bool:
+        """Determine if the entity is available."""
+        if self._data.client.advertisement_data_contains_state:
+            # If the device only uses advertisement data, just check if the address is present.
+            return async_address_present(self.hass, self._data.address)
+
+        # If the device needs an active connection, availability is based on whether it is connected.
+        return self._data.client.is_connected
+
+    @callback
+    def _handle_state_update(self, *args: Any) -> None:
+        """Handle state update."""
+        self.async_write_ha_state()
+
+    async def async_added_to_hass(self) -> None:
+        """Register callback."""
+        self.async_on_remove(
+            self._data.client.register_callback(self._handle_state_update)
+        )
+
+
+class EufyLifeRealTimeWeightSensorEntity(EufyLifeSensorEntity):
+    """Representation of an EufyLife real-time weight sensor."""
+
+    _attr_name = "Real-time weight"
+    _attr_native_unit_of_measurement = UnitOfMass.KILOGRAMS
+    _attr_device_class = SensorDeviceClass.WEIGHT
+
+    def __init__(self, data: EufyLifeData) -> None:
+        """Initialize the real-time weight sensor entity."""
+        super().__init__(data)
+        self._attr_unique_id = f"{data.address}_real_time_weight"
+
+    @property
+    def native_value(self) -> float | None:
+        """Return the native value."""
+        if self._data.client.state is not None:
+            return self._data.client.state.weight_kg
+        return None
```

Comment:
Maybe `self._data.client.state.realtime_weight_kg` so we have less logic to maintain here?

If it starts to get too long you could do this in the parent class

```
@property
def client_state(self):
   return self._data.client.state
```

Reply:
In my opinion this logic belongs in HA and not the library. In the bluetooth packets from the scale, there is just a weight reading along with a boolean (which I call `is_final`) indicating whether it is a finalized weight reading (occurs once the weight has stabilized and the scale "locks in" a reading).

The real time weight entity updates for all received packets, but the regular weight entity only updates when `is_final` is true. I think the separation of these two data points is a detail of the HA integration, since no such separation exists in the data itself. They are both reading from the same bytes in the data.

If you feel strongly about this then I'd be willing to change it. Just wanted to explain why I believed it should be this way.

Reply:
I think the value makes sense with or without HA involved. It doesn't seem like HA specific logic so I'd still put it in the library.

Reply:
OK, I moved the `is_final` check into the library. Now the library exposes two floats, `weight_kg` and `final_weight_kg` which are used directly by the integration.